### DISCUSSION TITLE
Pointcloud using Point Gaussians

### DISCUSSION THREAD

**Discussion by Findus23** [pyvista/pyvista#2576] [Q&A] [answered]
Hi,

First of all thanks for this amazing library! It is really easy to get great plots in a short amount of time.

There is just one thing I am missing to reproduce a visualisation from paraview:
I am trying to plot a large amount (>2M) points as simple points with low opacity and `point_size=1`. This works fine, but depends on being on exactly the right zoom level to be able to see the density of the points (as the point size seems to be independent of zoom).

In paraview a great way around this that also makes the visualisation better is using the "Point Gaussian" representation with "shader preset" gaussian blur instead.  

![grafik](https://user-images.githubusercontent.com/6266037/167587187-5eff44e8-a8a4-419c-8a3c-00178b544bf5.png)

Is there any way to get something similar with pyvista?

**Comment by banesullivan** [comment, marked as answer]
We don't (yet) support the points gaussian representation in PyVista - I'd love to see a PR to the end though.

One workaround is to manually perform the gaussian sampling of the points prior to plotting.

Could you perhaps provide this fantastic sample dataset for us to work on this challenge?

**Reply by akaszynski** [reply]
Let's open an issue on this. I'd like to see this as well.

**Reply by banesullivan** [reply]
See https://github.com/pyvista/pyvista/issues/2891 - didn't want to convert this to an issue.

**Reply by Findus23** [reply]
That's great to hear. If you want to use my datasets, I have exported a few of them here (as a CSV file with the coordinates as columns).
The particles correspond to (same mass) particles in a cosmological simulation and filtered to only show one resulting dark matter halo.
halo_low_res.csv has 32315 particles, halo_high_res.csv has 2062257 particles and halo_zoom.csv has 6656161 particles.

https://cloud.lw1.at/s/ZHt9GBww8QDHPps

**Reply by akaszynski** [reply]
Surprisingly easy!

![tmp](https://user-images.githubusercontent.com/11981631/176197093-6aff920b-b459-44d2-a99a-2a3a7f6f05f2.png)



```py
import pyvista
import numpy as np
import vtk

array = np.genfromtxt('/home/alex/Downloads/halo_low_res.csv', skip_header=1, delimiter=',')

pdata = pyvista.PolyData(array)

pl = pyvista.Plotter()
pl.add_points(pdata, point_size=0.1)
pl.camera.zoom(2)

# create a render pass
renderer = pl.renderer
basic_passes = vtk.vtkRenderStepsPass()

# blur the resulting image
# The blur delegates rendering the unblured image to the basic_passes
blur_pass = vtk.vtkGaussianBlurPass()
blur_pass.SetDelegatePass(basic_passes)

# tell the renderer to use our render pass pipeline
glrenderer = vtk.vtkOpenGLRenderer.SafeDownCast(renderer)
glrenderer.SetPass(blur_pass)

pl.show()

```

I'll have a PR for this before the day is out and include this in v0.35.1

**Reply by Findus23** [reply]
Many thanks for looking into this. This indeed also works for me and thanks to the blur makes things already quite a bit more visible.
But what really also be useful (but maybe an independent request) is another thing that the "Point Gaussian" representation does in paraview: The size of the dots (or gaussian blurs) becomes independent of the current zoom level, but is instead in "physical" units. This means you can zoom in and out into details without things becoming more or less visible due to distances between particles becoming larger when zooming in.

**Reply by akaszynski** [reply]
Got it. Do you know which vtk library they use?

**Reply by Findus23** [reply]
No, unfortunately I don't know that much about the vtk details apart from this quick summary of the feature:
https://www.kitware.com/paraview-5-introducing-the-point-gaussian-representation/
and my assumption that it is probably related to https://vtk.org/doc/nightly/html/classvtkPointGaussianMapper.html 

Independently, if the `PointGaussian` is really possible to access from pyvista, then the "sphere" visualisation would probably also be useful to many people for things like this (also from paraview): https://lw1.at/s/videos/simulations/one_more_collision.mp4

**Reply by banesullivan** [reply]
The API I envision for this will use "representation style", e.g.:

```py
p = pv.Plotter()
p.add_mesh(point_cloud, style='points_gaussian')
p.show()
```

I'd be happy to take on the implementation of this as I am somewhat familiar with how it works and have done some similar work in the past

**Reply by banesullivan** [reply]
Got it working, draft PR inbound!

![Screen Shot 2022-06-28 at 10 04 15 AM](https://user-images.githubusercontent.com/22067021/176227324-10cfae54-606d-445b-ae8b-7fa049c02a48.png)

**Reply by banesullivan** [reply]
See https://github.com/pyvista/pyvista/pull/2898

**Reply by kmarchais** [reply]
Hi, 
Thank you for this feature.
I would like to be able to scale point size according to the diameter of my spheres.
In paraview this option is available with "Use Scale Array" : [https://github.com/Kitware/ParaView/blob/master/Remoting/Views/vtkPointGaussianRepresentation.cxx](https://github.com/Kitware/ParaView/blob/master/Remoting/Views/vtkPointGaussianRepresentation.cxx)
Is there a way to implement this in PyVista if it's not already available?

**Reply by banesullivan** [reply]
Take a look at glyphing in PyVista: 

- https://docs.pyvista.org/examples/01-filter/glyphs.html
- https://docs.pyvista.org/examples/02-plot/movie_glyphs.html
- https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetFilters.glyph.html

**Reply by akaszynski** [reply]
BTW, point cloud gaussians is now implemented. See [Plotting Point Clouds](https://docs.pyvista.org/examples/02-plot/point-clouds.html#sphx-glr-examples-02-plot-point-clouds-py).

![image](https://user-images.githubusercontent.com/11981631/199784973-45596f5a-8ccc-4bb6-b504-9f9ecf63a771.png)

**Reply by kmarchais** [reply]
Thank you for your answers, I thought point gaussian representation would have better performance than glyphing.

**Reply by banesullivan** [reply]
It will have better performance yes - my bad. Though I'm not sure if the point size can be scaled by an array with it? Can it, @akaszynski ?
